controller/src/main/java/io/pravega/controller/server/rest/ModelHelper.java

```diff
@@ -60,15 +59,29 @@ public static final StreamConfiguration getCreateStreamConfig(final CreateStream
         if (createStreamRequest.getRetentionPolicy() != null) {
             switch (createStreamRequest.getRetentionPolicy().getType()) {
                 case LIMITED_SIZE_MB:
-                    retentionPolicy = RetentionPolicy.bySizeBytes(
-                            createStreamRequest.getRetentionPolicy().getValue() * 1024 * 1024);
+                    if (createStreamRequest.getRetentionPolicy().getMaxValue() == 0) {
+                        // max value is not specified
+                        retentionPolicy = RetentionPolicy.bySizeBytes(
+                                createStreamRequest.getRetentionPolicy().getValue() * MB_TO_BYTES);
+                    } else {
+                        retentionPolicy = RetentionPolicy.bySizeBytes(
+                                createStreamRequest.getRetentionPolicy().getValue() * MB_TO_BYTES,
+                                createStreamRequest.getRetentionPolicy().getMaxValue() * MB_TO_BYTES);
+                    }
                     break;
                 case LIMITED_DAYS:
-                    retentionPolicy = getRetentionPolicy(createStreamRequest.getRetentionPolicy().getTimeBasedRetention(),
-                            createStreamRequest.getRetentionPolicy().getValue());
-                    break;
-                case CONSUMPTION:
-                    retentionPolicy = getConsumptionRetentionPolicy(createStreamRequest.getRetentionPolicy().getConsumptionLimits());
+                    if (createStreamRequest.getRetentionPolicy().getMaxValue() == 0
+                        && createStreamRequest.getRetentionPolicy().getMaxTimeBasedRetention().getDays() == 0
```

Comment:
shouldnt this check simply be:
```
createStreamRequest.getRetentionPolicy().getMaxTimeBasedRetention() != null)
```
otherwise this access attempt itself can throw NPE.
Note: this is json derialization happening here, so the maxTimeBasedRetention will be defaulted to null if its not specified in the json payload.

Reply:
Fixed & also added tests to cover max Limit validation